### ISSUE TITLE
LXD-created LXC container is detected as a Xen domU

### ISSUE THREAD

**Issue by kfix** [saltstack/salt#43405]
### Setup
LXD Container host:
* Ubuntu 16.04 w/HWE kernel 
```
# uname -a
Linux localhost 4.8.0-56-generic #61~16.04.1-Ubuntu SMP Wed Jun 14 11:58:22 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

$ dpkg-query -W linux-image-generic-hwe-16.04
linux-image-generic-hwe-16.04	4.10.0.33.35

$ tree /sys/bus/xen
/sys/bus/xen
├── devices
├── drivers
├── drivers_autoprobe
├── drivers_probe
└── uevent

$ ls /proc/sys
abi  debug  dev  fs  kernel  net  vm

$ ls /proc/bus
input  pc
```

LXD container:
* Ubuntu 14.04 w/salt installed from mirror of saltstack deb repo
* /sys, /proc/sys, /proc/bus are bind-mounted from host 
* no states and pillars are applied

### Steps to Reproduce Issue
```
# virt-what
lxc

# salt-call --local grains.get virtual
local:
    xen

# salt-call --local grains.get virtual_subtype
local:
    Xen PV DomU
```
### Versions Report
```# salt-call --versions-report
Salt Version:
           Salt: 2016.3.4

Dependency Versions:
           cffi: 1.5.2
       cherrypy: Not Installed
       dateutil: 2.4.2
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: 0.24.0
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
         pygit2: 0.24.0
         Python: 2.7.12 (default, Nov 19 2016, 06:48:10)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: Ubuntu 16.04 xenial
        machine: x86_64
        release: 4.8.0-56-generic
         system: Linux
        version: Ubuntu 16.04 xenial
```

The problematic detection is here:
https://github.com/saltstack/salt/blob/237e5ce697895844ba0c8b0deaecbc4e37afb5b6/salt/grains/core.py#L781

that test is tripped by this randomly named veth on my *host* machine (containers share a ring buffer).

```
# dmesg | grep -i xen:
[4650609.480364] IPv6: ADDRCONF(NETDEV_UP): vethSV3XEN: link is not ready
[4650609.624532] IPv6: ADDRCONF(NETDEV_CHANGE): vethSV3XEN: link becomes ready
```

Also, it is an lxc container created by lxd, so it has a /dev/lxd/sock file.
So maybe the grains should be:
```
virtual: lxc
virtual_subtype: lxd
```

**Comment by kfix** [comment]
running `sudo sysctl -w kernel.dmesg_restrict=1` on the host will block non-host-root users from accessing the log-ring and `virtual: lxc` is detected again. but that is a drastic move to prevent this misunderstanding.

**Comment by gtmanfred** [comment]
I guess you are running on a virtual machine that is running on a xen server?

Because the only way it would get there was if the lxc container had 

```
        if isdir('/proc/sys/xen') or \
                isdir('/sys/bus/xen') or isdir('/proc/xen'):
```

```
diff --git a/salt/grains/core.py b/salt/grains/core.py
index bb1696304a..acd20e109b 100644
--- a/salt/grains/core.py
+++ b/salt/grains/core.py
@@ -694,17 +694,6 @@ def _virtual(osdata):
                     grains['virtual_subtype'] = 'chroot'
             except (IOError, OSError):
                 pass
-        if os.path.isfile('/proc/1/cgroup'):
-            try:
-                with salt.utils.fopen('/proc/1/cgroup', 'r') as fhr:
-                    if ':/lxc/' in fhr.read():
-                        grains['virtual_subtype'] = 'LXC'
-                with salt.utils.fopen('/proc/1/cgroup', 'r') as fhr:
-                    fhr_contents = fhr.read()
-                    if ':/docker/' in fhr_contents or ':/system.slice/docker' in fhr_contents:
-                        grains['virtual_subtype'] = 'Docker'
-            except IOError:
-                pass
         if isdir('/proc/vz'):
             if os.path.isfile('/proc/vz/version'):
                 grains['virtual'] = 'openvzhn'
@@ -776,6 +765,17 @@ def _virtual(osdata):
                         grains['virtual'] = 'bhyve'
             except IOError:
                 pass
+        if os.path.isfile('/proc/1/cgroup'):
+            try:
+                with salt.utils.fopen('/proc/1/cgroup', 'r') as fhr:
+                    if ':/lxc/' in fhr.read():
+                        grains['virtual_subtype'] = 'LXC'
+                with salt.utils.fopen('/proc/1/cgroup', 'r') as fhr:
+                    fhr_contents = fhr.read()
+                    if ':/docker/' in fhr_contents or ':/system.slice/docker' in fhr_contents:
+                        grains['virtual_subtype'] = 'Docker'
+            except IOError:
+                pass
     elif osdata['kernel'] == 'FreeBSD':
         kenv = salt.utils.which('kenv')
         if kenv:

```

you could try applying something like this and moving the lxc detection to the end?  since it seems like it is getting overwritten by the xen virtual subtype.

If that works we can submit a patch and try and make sure it doesn't break anything else?

Thanks,
Daniel

**Comment by kfix** [comment]
I updated the report-comment with more details.
The HWE kernel has a built in DomU driver that I can't unload, but this kernel & host OS is running on a bare metal machine.

**Comment by gtmanfred** [comment]
Did you try moving the LXC discovery to later in the salt.grains.core file and see if that will solve the issue?

If it does we would appreciate a PR.

Thanks,
Daniel

**Comment by kfix** [comment]
sure, I'll install an updated minion and give that a try

**Comment by gtmanfred** [comment]
An easier way would be to drop the core.py file in `salt://_grains` and run `salt \* saltutil.sync_grains`

https://docs.saltstack.com/en/latest/ref/file_server/dynamic-modules.html

**Comment by nages13** [comment]
Moved down the code after xen discovery and submitted PR #46756.

**Comment by rallytime** [comment]
Closed via #46756
